os/kernel/debug/mem_leak_checker.c

```diff
@@ -0,0 +1,437 @@
+#include <sys/types.h>
+#include <tinyara/mm/mm.h>
+#include <tinyara/mm/heap_regioninfo.h>
+#include <binary_manager/binary_manager.h>
+
+#include "../binfmt/libelf/libelf.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define MEM_USED   0
+#define MEM_LEAK   1
+#define MEM_BROKEN 2
+
+#define MEM_ACCESS_UNIT    0x04
+#define MAX_ALLOC_COUNT    CONFIG_MEM_LEAK_CHECKER_MAX_ALLOC_COUNT
+#define HASH_SIZE          CONFIG_MEM_LEAK_CHECKER_HASH_TABLE_SIZE
+
+#define RAM_START_ADDR  kregionx_start[0]
+#define RAM_END_ADDR	(kregionx_start[CONFIG_KMM_REGIONS - 1] + kregionx_size[CONFIG_KMM_REGIONS - 1])
+#define HEAP_START_ADDR leak_checker.heap_start
+#define HEAP_END_ADDR   leak_checker.heap_end
+
+struct alloc_node_info_s {
+	volatile struct mm_allocnode_s *node;
+	struct alloc_node_info_s *next;
+};
+
+static struct alloc_node_info_s *hash_table[HASH_SIZE];
+static struct alloc_node_info_s node_info[MAX_ALLOC_COUNT];
+static struct mem_leak_checker_info_s leak_checker;
+
+#ifdef CONFIG_MM_KERNEL_HEAP
+extern struct mm_heap_s g_kmmheap[CONFIG_KMM_NHEAPS];
+#endif
+extern sq_queue_t g_bin_addr_list;
+static void hash_init(void)
+{
+	int index;
+	for (index = 0; index < HASH_SIZE; ++index) {
+		hash_table[index] = NULL;
+	}
+}
+
+static void add_hash(int index)
+{
+	long key;
+	struct alloc_node_info_s *cur;
+	key = (long)node_info[index].node % HASH_SIZE;
+	if (hash_table[key] == NULL) {
+		hash_table[key] = &node_info[index];
+		return;
+	}
+
+	cur = hash_table[key];
+	while (cur->next) {
+		cur = cur->next;
+	}
+	cur->next = &node_info[index];
+}
+
+static bool search_hash(unsigned long value)
+{
+	long key = value % HASH_SIZE;
+	struct alloc_node_info_s *cur = hash_table[key];
+	while (cur != NULL) {
+		if ((unsigned long)cur->node == value && (unsigned long)cur->node != (value + (unsigned long)SIZEOF_MM_ALLOCNODE)) {
+			if (cur->node->reserved == MEM_USED) {
+				return false;
+			}
+			cur->node->reserved = MEM_USED;
+			return true;
+		}
+		if (cur->next == NULL) {
+			return false;
+		}
+		cur = cur->next;
+	}
+	return false;
+}
+
+static int get_node_cnt(void)
+{
+	volatile struct mm_allocnode_s *node;
+	mmsize_t node_size;
+	node_size = SIZEOF_MM_ALLOCNODE;
+
+	int ret = 0;
+	
+#if CONFIG_KMM_REGIONS > 1
+	int region;
+#else
+#define region 0
+#endif
+
+	/* Visit each region */
+
+#if CONFIG_KMM_REGIONS > 1
+	for (region = 0; region < leak_checker.regions; region++)
+#endif
+	{
+		node_size = SIZEOF_MM_ALLOCNODE;
+		for (node = leak_checker.heap_start[region]; node < leak_checker.heap_end[region]; node = (struct mm_allocnode_s *)((char *)node + node->size)) {
+			/* Ignore the heap start checking, because there is a guard node in heap start */
+			if (node == leak_checker.heap_start[region]) {
+				continue;
+			}
+			/* Check broken link */
+			if (node_size != (node->preceding & 0x7FFFFFFF)) {
```

Comment:
As a follow-up to our discussion last time, how about adding a macro for obtaining the previous node size?

https://github.com/Samsung/TizenRT/blob/022650271860ce20551146c7cde543904d21ed3d/os/mm/mm_heap/mm_check_heap_corruption.c#L67